### PR TITLE
Add documentation versioning

### PR DESCRIPTION
**Pull request by samijaber**
### Screenshot
screen recording 2017-04-22 at 09 38 pm

### Overview

- improves `build-docs` script to include the versioning of documentation, with `edge` being the cutting-edge, yet-unreleased code documentation.
- adds a `VersionSelector` dropdown menu to allow users to choose the documentation version.
- provide the documentation JSON as a navigation parameter in the website.
- retroactively adds documentation for `1.0.0-beta.7` and `1.0.0-beta.8`. I downloaded the zipped release for `1.0.0-beta.7` and used the `docs-dist.json` that exists there. As for `1.0.0-beta.8`, I checked out this commit e94c9c1061151a1478b47640710bcafdbe738848 and took the documentation from there.
- fixes https://github.com/react-community/react-navigation/issues/1060

Now, when `npm run build-docs` (or `build` or `prepublish`) are executed, the `compile-docs` script will do three things:

- First, the current `edge.json` is renamed to the current version (based on whatever version is in `package.json`). If a file for the current version already exists, it will overwrite it.
- Second, a new `edge.json` is built (this step is what's already there)
- Last, the `docs/index.js` file is rebuilt, in case the 1st step resulted in a new file being created

### To-do

- [ ] improve styling of `VersionSelector` (I wanted to make sure this PR is created first so that I can get feedback on the script while I tinker with the styling).

### CODE REVIEW

**Comment by samijaber** [comment]
cc @grabbou : you mentioned trying to solve this in #1060. Take a look at this PR and let me know if it already does the trick! 🙂

**Comment by grabbou** [comment]
Hey, thanks for taking a stub on this! Overall, looking good, I am leaving few comments here and there.

File: website/src/AppFrame.js [hunk 1]
```diff
@@ -36,26 +43,105 @@ const NavigationLinks = ({navigation, className, reverse}) => {
...
+  }
+
+  toggleMenuShown() {
+    this.setState({ menuShown: !this.state.menuShown });
+  }
+
+  handleVersionClick(version) {
+    const { setParams } = this.props.childNavigation;
+    return () => {
+      setParams({ version });
```
Comment (grabbou) [inline website/src/AppFrame.js]: We shouldn't return a closure here, see comment below where we call onPress.

File: website/src/AppFrame.js [hunk 2]
```diff
@@ -36,26 +43,105 @@ const NavigationLinks = ({navigation, className, reverse}) => {
...
+      isUndef(state.params.version)
+    );
+
+    const currentVersion = docVersionIsUndefined(state) ? 'edge' : state.params.version;
+    const versionsList = Object.keys(docs).map((version) => (
+      <li
+        key={version}
+        onClick={this.handleVersionClick(version)}
+        style={{
+          color: (currentVersion === version) ? '#6b52ae' : '#555'
```
Comment (grabbou) [inline website/src/AppFrame.js]: nit: `( )` seem unnecessary?

File: website/src/AppFrame.js [hunk 3]
```diff
@@ -36,26 +43,105 @@ const NavigationLinks = ({navigation, className, reverse}) => {
...
+    ));
+
+    return (
+      <div className='versionselector'>
+        <span className="versionbutton" onClick={this.toggleMenuShown}>
+          version: {currentVersion} ▼
+        </span>
+        <ul
+          style={{
+            //display: menuShown ? 'inline' : 'none',
```
Comment (grabbou) [inline website/src/AppFrame.js]: nit

File: website/src/AppFrame.js [hunk 4]
```diff
@@ -36,26 +43,105 @@ const NavigationLinks = ({navigation, className, reverse}) => {
...
 class AppFrame extends React.Component {
-  state = {isMobileMenuOpen: false};
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      isMobileMenuOpen: false,
+    };
+
+    this.setVersion = this.setVersion.bind(this);
```
Comment (grabbou) [inline website/src/AppFrame.js]:
```
setVersion = (version) => {

}
```
and we don't have to `bind`?

File: website/src/AppFrame.js [hunk 5]
```diff
@@ -36,26 +43,105 @@ const NavigationLinks = ({navigation, className, reverse}) => {
...
+  
   componentWillReceiveProps(props) {
     if (this.props.navigation.state !== props.navigation.state) {
       this.setState({isMobileMenuOpen: false});
       window.scrollTo(0,0);
     }
   }
+  
+  setVersion(version) {
+    return () => this.setState({ version });
```
Comment (grabbou) [inline website/src/AppFrame.js]: Same as previously, no need to return function here. Prefer `onPress={() => setVersion(version)}` instead

File: website/src/MDPage.js [hunk 6]
```diff
@@ -29,9 +27,20 @@ const getHeadingForLevel = (level) => {
   }
 };
 
+const isDefined = (obj) => obj !== undefined;
+
+const getDocsMD = (docPath, navigation) => {
+  const { state } = navigation;
+  const docVersion = (isDefined(state) && isDefined(state.params))
```
Comment (grabbou) [inline website/src/MDPage.js]: AFAIK `navigation.state` is always defined. We can only check for `state.params && state.params.version`

File: website/src/MDPage.js [hunk 7]
```diff
@@ -29,9 +27,20 @@ const getHeadingForLevel = (level) => {
...
+    ? state.params.version
+    : 'edge';
```
Comment (grabbou) [inline website/src/MDPage.js]: That logic is already defined in two places - here and in version selector. We could either use `initialRouteParams` for cases when params are not set?

**Comment by ericvicenti** [comment]
Could we possibly change the design to integrate the dropdown into the "docs" link, rather than put this in the center of the header? When the docs link is inactive, it can mention the version, like "DOCS - 1.0.0-beta.8", and when it is active, it has the dropdown functionality.

Also I notice this does not integrate with the path, so deep links may be broken if you attempt to link to an old version of the docs.

Be sure to test this on mobile, and post a screenshot of that behavior.

**Comment by grabbou** [comment]
@samijaber will you have time soon to work on this? If not, let me know and I'll be happy to take it. We need working switch for versions soon as we keep development more and more active!

**Comment by samijaber** [comment]
@grabbou I'm unfortunately somewhat busy over the next few days. Feel free to take over and make the changes you want! Otherwise, I might be able to take a look in a week or so.

**Comment by kelset** [comment]
Hey @grabbou @samijaber is this PR worth to keep open? No updates since April and the issue mentioned #1060 is closed atm.

**Comment by brentvatne** [comment]
wow, thank you so much for doing this, sincere apologies that nobody was able to land it. we're in the process of moving over to using docusaurus for our website/docs, so this PR is no longer relevant, you can follow along at http://github.com/react-navigation/website

**Pull request closed without merging**